httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/gss/GssSchemeBase.java

```diff
@@ -0,0 +1,394 @@
+    private static final String KERBEROS_SCHEME = "HTTP";
+
+    // The GSS spec does not specify how long the conversation can be. This should be plenty.
+    // Realistically, we get one initial token, then one maybe one more for mutual authentication.
+    private static final int MAX_GSS_CHALLENGES = 3;
+    private final GssConfig config;
+    private final DnsResolver dnsResolver;
+    private final boolean requireMutualAuth;
+    private final boolean ignoreMissingToken;
+    private int challengesLeft = MAX_GSS_CHALLENGES;
+
+    /** Authentication process state */
+    private State state;
+    private GSSCredential gssCredential;
+    private GSSContext gssContext;
+    private String challenge;
+    private byte[] queuedToken = new byte[0];
+
+    GssSchemeBase(final GssConfig config, final DnsResolver dnsResolver) {
+        super();
+        this.config = config != null ? config : GssConfig.DEFAULT;
+        this.dnsResolver = dnsResolver != null ? dnsResolver : SystemDefaultDnsResolver.INSTANCE;
+        this.requireMutualAuth = config.isRequireMutualAuth();
+        this.ignoreMissingToken = config.isIgnoreMissingToken();
+        this.state = State.UNINITIATED;
+    }
+
+    GssSchemeBase(final GssConfig config) {
+        this(config, SystemDefaultDnsResolver.INSTANCE);
+    }
+
+    GssSchemeBase() {
+        this(GssConfig.DEFAULT, SystemDefaultDnsResolver.INSTANCE);
+    }
+
+    @Override
+    public String getRealm() {
+        return null;
+    }
+
+    // Required by AuthScheme for backwards compatibility
+    @Override
+    public void processChallenge(final AuthChallenge authChallenge,
+            final HttpContext context ) {
+        // If this gets called, then AuthScheme was changed in an incompatible way
+        throw new UnsupportedOperationException();
+    }
+
+    // The AuthScheme API maps awkwardly to GSSAPI, where proccessChallange and generateAuthResponse
+    // map to the same single method call. Hence the generated token is only stored in this method.
+    @Override
+    public void processChallenge(
+            final HttpHost host,
+            final boolean challenged,
+            final AuthChallenge authChallenge,
+            final HttpContext context
+            ) throws AuthenticationException {
+
+        if (challengesLeft-- <= 0 ) {
+            if (LOG.isDebugEnabled()) {
+                final HttpClientContext clientContext = HttpClientContext.cast(context);
+                final String exchangeId = clientContext.getExchangeId();
+                LOG.debug("{} GSS error: too many challenges received. Infinite loop ?", exchangeId);
+            }
+            // TODO: Should we throw an exception ? There is a test for this behaviour.
+            state = State.FAILED;
+            return;
+        }
+
+        final byte[] challengeToken = Base64.decodeBase64(authChallenge == null ? null : authChallenge.getValue());
```

Comment:
Why bother Base64 if we can do the ternary check outside?

Reply:
I'm not sure what you mean. 
The ternary here prevents an NPE from authChallenge.getValue() is authChallange is null.
Base64.decodeBase64 returns null for null, so we don't have to check for a null value.

Reply:
`authChallenge == null ? null : Base64.decodeBase64(authChallenge.getValue());`